std-bits/src/main/java/org/enso/base/json/Parser.java

```diff
@@ -0,0 +1,449 @@
+        throw new UnexpectedTokenException(position, "a comma or a closing bracket");
+    }
+  }
+
+  private static int consumeAny(
+      char[] chars, int position, JsonConsumer consumer, Deque<State> state) {
+    switch (chars[position]) {
+      case '[':
+        consumer.on_start_array();
+        position++;
+        state.push(State.ARRAY_END_OR_VALUE);
+        return position;
+      case '{':
+        consumer.on_start_object();
+        position++;
+        state.push(State.OBJECT_KEY_OR_END);
+        return position;
+      case '"':
+        return consumeString(chars, position, consumer, false);
+      case '-':
+      case '0':
+      case '1':
+      case '2':
+      case '3':
+      case '4':
+      case '5':
+      case '6':
+      case '7':
+      case '8':
+      case '9':
+        return consumeNumber(chars, position, consumer);
+      case 'n':
+        return consumeNull(chars, position, consumer);
+      case 't':
+        return consumeTrue(chars, position, consumer);
+      case 'f':
+        return consumeFalse(chars, position, consumer);
+    }
+    throw new UnexpectedTokenException(position, "a start of a JSON value");
+  }
+
+  private static int consumeNull(char[] chars, int position, JsonConsumer consumer) {
+    if (position + 3 < chars.length) {
+      boolean match =
+          chars[position] == 'n'
+              && chars[position + 1] == 'u'
+              && chars[position + 2] == 'l'
+              && chars[position + 3] == 'l';
+      if (match) {
+        consumer.on_null();
+        return position + 4;
+      }
+      throw new UnexpectedTokenException(position, "a null");
+    }
+    throw new UnexpectedEndOfInputException("a null");
+  }
+
+  private static int consumeTrue(char[] chars, int position, JsonConsumer consumer) {
+    if (position + 3 < chars.length) {
+      boolean match =
+          chars[position] == 't'
+              && chars[position + 1] == 'r'
+              && chars[position + 2] == 'u'
+              && chars[position + 3] == 'e';
+      if (match) {
+        consumer.on_true();
+        return position + 4;
+      }
+      throw new UnexpectedTokenException(position, "a true");
+    }
+    throw new UnexpectedEndOfInputException("a true");
+  }
+
+  private static int consumeFalse(char[] chars, int position, JsonConsumer consumer) {
+    if (position + 4 < chars.length) {
+      boolean match =
+          chars[position] == 'f'
+              && chars[position + 1] == 'a'
+              && chars[position + 2] == 'l'
+              && chars[position + 3] == 's'
+              && chars[position + 4] == 'e';
+      if (match) {
+        consumer.on_false();
+        return position + 5;
+      }
+      throw new UnexpectedTokenException(position, "a false");
+    }
+    throw new UnexpectedEndOfInputException("a false");
+  }
+
+  private static int consumeString(
+      char[] chars, int position, JsonConsumer consumer, boolean isKey) {
+    if (chars[position] != '\"') {
```

Comment:
Here you escape the `"` inside of char quotes `' '` but a few lines below you do not.